src/TestFramework/MSTest.Core/Attributes/DynamicDataAttribute.cs

```diff
@@ -0,0 +1,132 @@
+        /// Data is declared as property.
+        /// </summary>
+        Property = 0,
+
+        /// <summary>
+        /// Data is declared in method.
+        /// </summary>
+        Method = 1
+    }
+
+    /// <summary>
+    /// Attribute to define dynamic data for a test method.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
+    public class DynamicDataAttribute : TestDataSource
+    {
+        private string dynamicDataSourceName;
+
+        private Type dynamicDataDeclaringType;
+
+        private DynamicDataSourceType dynamicDataSourceType;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DynamicDataAttribute"/> class.
+        /// </summary>
+        /// <param name="dynamicDataSourceName">
+        /// The name of method or property having test data.
+        /// </param>
+        /// <param name="dynamicDataSourceType">
+        /// Specifies whether the data is stored as property or in method.
+        /// </param>
+        public DynamicDataAttribute(string dynamicDataSourceName, DynamicDataSourceType dynamicDataSourceType = DynamicDataSourceType.Property)
+        {
+            this.dynamicDataSourceName = dynamicDataSourceName;
+            this.dynamicDataSourceType = dynamicDataSourceType;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DynamicDataAttribute"/> class.
+        /// </summary>
+        /// <param name="dynamicDataSourceName">
+        /// The type in which the test data is declared (as property or in method).
+        /// </param>
+        /// <param name="dynamicDataDeclaringType">
+        /// The declaring type of property or method having data.
+        /// </param>
+        /// <param name="dynamicDataSourceType">
+        /// Specifies whether the data is stored as property or in method.
+        /// </param>
+        public DynamicDataAttribute(string dynamicDataSourceName, Type dynamicDataDeclaringType, DynamicDataSourceType dynamicDataSourceType = DynamicDataSourceType.Property)
+        {
+            this.dynamicDataSourceName = dynamicDataSourceName;
+            this.dynamicDataDeclaringType = dynamicDataDeclaringType;
+            this.dynamicDataSourceType = dynamicDataSourceType;
+        }
+
+        /// <inheritdoc />
+        public override IEnumerable<object[]> GetData(MethodInfo methodInfo)
+        {
+            // Check if the declaring type of test data is passed in constructor. If not, default to test method's class type.
+            if (this.dynamicDataDeclaringType == null)
+            {
+                this.dynamicDataDeclaringType = methodInfo.DeclaringType;
+            }
+
+            object obj = null;
+
+            switch (this.dynamicDataSourceType)
+            {
+                case DynamicDataSourceType.Property:
+                    var property = this.dynamicDataDeclaringType.GetTypeInfo().GetDeclaredProperty(this.dynamicDataSourceName);
```

Comment:
would this be GetRunTimeProperty instead allowing properties defined in the base class?